wasm/jit/jit_amd64.go

```diff
@@ -161,18 +161,22 @@ func (c *amd64Compiler) generate() ([]byte, uint64, error) {
 	if err != nil {
 		return nil, 0, err
 	}
-	// As we cannot read RIP register directly,
-	// we calculate now the offset to the next instruction
-	// relative to the beginning of this function body.
+	// As we cannot read RIP register directly, we calculate now the offset to the next
+	// instruction after return instruction for function call.
 	const operandSizeBytes = 8
-	for _, obj := range c.requireFunctionCallReturnAddressOffsetResolution {
+	for _, inst := range c.requireFunctionCallReturnAddressOffsetResolution {
+		afterReturnInst := inst
+		// We move forward on the linked list of instructions until the return instrucion found.
+		for ; ; afterReturnInst = afterReturnInst.Link {
+			if afterReturnInst.As == obj.ARET {
+				// Now we found the return instruction, move forward once again.
+				afterReturnInst = afterReturnInst.Link
+				break
+			}
+		}
 		// Skip MOV, and the register(rax): "0x49, 0xbd"
-		start := obj.Pc + 2
-		// obj.Link = setting offset to memory
-		// obj.Link.Link = writing back the stack pointer to eng.stackPointer.
-		// obj.Link.Link.Link = Return instruction.
-		// Therefore obj.Link.Link.Link.Link means the next instruction after the return.
-		afterReturnInst := obj.Link.Link.Link.Link
```

Comment:
This was fragile as the assembler sometimes automatically inserts NOP instruction to make all the instructions aligned

Reply:
and actually turned out to be a bug caught by spectest!